### PR TITLE
Change to StoreCodeUnchecked in Genesis and snapshotter

### PR DESCRIPTION
**Pull request by GNaD13**
## Description

closes: #5155 

### Commit Message / Changelog Entry

```text
chore: use StoreCodeUnchecked in genesis and state sync.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go [hunk 1]
```diff
@@ -15,6 +15,13 @@ type WasmEngine interface {
 	// It does the same as StoreCodeUnchecked plus the static checks.
 	StoreCode(code wasmvm.WasmCode) (wasmvm.Checksum, error)
 
+	// Create will compile the wasm code, and store the resulting pre-compile
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:18]:
```suggestion
	// StoreCodeUnchecked will compile the wasm code, and store the resulting pre-compile
```

File: modules/light-clients/08-wasm/keeper/keeper.go [hunk 2]
```diff
@@ -118,7 +118,7 @@ func (Keeper) storeWasmCode(ctx sdk.Context, code []byte) ([]byte, error) {
 
 	// create the code in the vm
 	ctx.GasMeter().ConsumeGas(types.VMGasRegister.CompileCosts(len(code)), "Compiling wasm bytecode")
-	vmChecksum, err := ibcwasm.GetVM().StoreCode(code)
+	vmChecksum, err := ibcwasm.GetVM().StoreCodeUnchecked(code)
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/keeper/keeper.go:121]: I think we need to do either `StoreCode` or `StoreCodeUnchecked` depending on where this keeper `storeWasmCode` is being called from. There are two situations:

- keeper `storeWasmCode` is [called in `InitGenesis`](https://github.com/cosmos/ibc-go/blob/2bd29c08fd1fe50b461fc33a25735aa792dc896e/modules/light-clients/08-wasm/keeper/genesis.go#L14) -> We should call `StoreCodeUnchecked`.
- keeper `storeWasmCode` is [called in `StoreCode` RPC handler](https://github.com/cosmos/ibc-go/blob/2bd29c08fd1fe50b461fc33a25735aa792dc896e/modules/light-clients/08-wasm/keeper/msg_server.go#L25) -> We should call `StoreCode`.

**Review by crodriguezvega — COMMENTED** [review]
Thanks for working on this one, @GNaD13! I left a comment explaining that we should be more granular on what function is called depending the call site of keeper `storeWasmCode`. I hope it makes sense, but feel free to ask questions, if you have any!

**Comment by GNaD13** [comment]
> Thanks for working on this one, @GNaD13! I left a comment explaining that we should be more granular on what function is called depending the call site of keeper `storeWasmCode`. I hope it makes sense, but feel free to ask questions, if you have any!

Hi @crodriguezvega, thanks for your reviewed. I've updated the code.

**Comment by GNaD13** [comment]
> Thank you very much for this one, @GNaD13! I just pushed a quick update, because I thought of a way to avoid duplicating the code of `storeWasmCode` just to have one line different.

thank you sir @crodriguezvega

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go [hunk 3]
```diff
@@ -8,13 +8,20 @@ import (
...
 type WasmEngine interface {
-	// Create will compile the wasm code, and store the resulting pre-compile
+	// StoreCode will compile the wasm code, and store the resulting pre-compile
 	// as well as the original code. Both can be referenced later via checksum
 	// This must be done one time for given code, after which it can be
 	// instatitated many times, and each instance called many times.
 	// It does the same as StoreCodeUnchecked plus the static checks.
 	StoreCode(code wasmvm.WasmCode) (wasmvm.Checksum, error)
 
+	// StoreCodeUnchecked will compile the wasm code, and store the resulting pre-compile
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:18]: The godoc for this function is identical to that above. Could we add more clarity to differences between `StoreCode` and `StoreCodeUnchecked`

Comment (damiannolan) [reply modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:18]: Oh.. whoops! Yeah its not super obvious. Any idea what exactly the static checks are? Maybe we could add a brief description on that.

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go [hunk 4]
```diff
@@ -8,13 +8,20 @@ import (
...
+	// as well as the original code. Both can be referenced later via checksum
+	// This must be done one time for given code, after which it can be
+	// instatitated many times, and each instance called many times.
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:21]:
```suggestion
	// instantitated many times, and each instance called many times.
```

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go [hunk 5]
```diff
@@ -8,13 +8,20 @@ import (
 var _ WasmEngine = (*wasmvm.VM)(nil)
 
 type WasmEngine interface {
-	// Create will compile the wasm code, and store the resulting pre-compile
+	// StoreCode will compile the wasm code, and store the resulting pre-compile
 	// as well as the original code. Both can be referenced later via checksum
 	// This must be done one time for given code, after which it can be
 	// instatitated many times, and each instance called many times.
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:14]:
```suggestion
	// instantiated many times, and each instance called many times.
```

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go [hunk 6]
```diff
@@ -8,13 +8,20 @@ import (
...
+	// instantiated many times, and each instance called many times.
 	// It does the same as StoreCodeUnchecked plus the static checks.
 	StoreCode(code wasmvm.WasmCode) (wasmvm.Checksum, error)
 
+	// StoreCodeUnchecked will compile the wasm code, and store the resulting pre-compile
+	// as well as the original code. Both can be referenced later via checksum
+	// This must be done one time for given code, after which it can be
+	// instantiated many times, and each instance called many times.
+	// It does the same as StoreCode but without the static checks.
+	StoreCodeUnchecked(code wasmvm.WasmCode) (wasmvm.Checksum, error)
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:23]: what about this addition? @DimitrisJim 

```suggestion
         // This allows restoring previous contract code in genesis and state-sync that may have been initially stored under different configuration constraints.
	StoreCodeUnchecked(code wasmvm.WasmCode) (wasmvm.Checksum, error)
```

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go (hunk 3 above)
Comment (damiannolan) [reply modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:18]: Added a suggestion in another comment

File: modules/light-clients/08-wasm/keeper/keeper.go [hunk 7]
```diff
@@ -91,7 +91,7 @@ func (k Keeper) GetAuthority() string {
 	return k.authority
 }
 
-func (Keeper) storeWasmCode(ctx sdk.Context, code []byte) ([]byte, error) {
+func (Keeper) storeWasmCode(ctx sdk.Context, code []byte, storeFn func(code wasmvm.WasmCode) (wasmvm.Checksum, error)) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/keeper.go:94]: Would support the idea of an alias 👍🏻

**Pull request merged**